File: testpaper/extract_datajs.py
#!/usr/bin/env python3
"""
extract_datajs.py  (v2 - fast regex-based approach)
====================================================
Reads data.js, converts JS object syntax to valid JSON using
regex-based string-aware key quoting (no per-character Python loops),
then writes one JSON file per subject to static-data/.

Why v2: v1 used character-by-character append → MemoryError on 14 MB.
v2 uses re.finditer to split at string boundaries, applies re.sub
only to non-string segments. The C-level regex engine handles 14 MB fast.

Usage:  py -3 extract_datajs.py
Output: static-data/<subject>.json
        static-data/manifest.json
        extract_datajs.log
"""

import os
import re
import sys
import json
import traceback

DATA_FILE  = "data.js"
OUTPUT_DIR = "static-data"
LOG_FILE   = "extract_datajs.log"

log_lines = []

def log(msg):
    print(msg, flush=True)
    log_lines.append(msg)

def write_log():
    try:
        with open(LOG_FILE, "w", encoding="utf-8") as f:
            f.write("\n".join(log_lines) + "\n")
    except Exception as e:
        print(f"[WARN] Could not write log: {e}", flush=True)


# ─── Fast JS → JSON conversion ────────────────────────────────────────────────
# Strategy:
#   1. Split content into (non-string, string) alternating chunks using regex
#   2. In non-string chunks: quote unquoted identifier keys
#   3. String chunks: pass through untouched
#   4. After reassembly: strip trailing commas (valid JS, invalid JSON)

# Matches any double-quoted JS string including escape sequences
_STRING_RE = re.compile(r'"(?:[^"\\]|\\.)*"', re.DOTALL)

# Matches an unquoted JS identifier used as an object key:
#   word boundary + identifier chars + optional whitespace + colon (not ::)
_KEY_RE = re.compile(r'(?<!["\w])([a-zA-Z_$][a-zA-Z0-9_$]*)(\s*:)(?!:)')


def _quote_keys_in_chunk(chunk):
    """Quote all bare identifier keys in a non-string JS chunk."""
    return _KEY_RE.sub(lambda m: f'"{m.group(1)}"' + m.group(2), chunk)


def js_to_json(js_str):
    """
    Convert a JS object literal to valid JSON.
    - Quotes unquoted identifier keys (only in non-string regions)
    - Strips trailing commas before ] and }
    - Replaces `undefined` with `null` in non-string regions
    """
    parts = []
    prev_end = 0

    for m in _STRING_RE.finditer(js_str):
        start, end = m.span()
        # Non-string chunk before this string
        chunk = js_str[prev_end:start]
        chunk = _quote_keys_in_chunk(chunk)
        chunk = chunk.replace('undefined', 'null')
        parts.append(chunk)
        # String: pass through verbatim
        parts.append(m.group(0))
        prev_end = end

    # Final non-string tail
    tail = js_str[prev_end:]
    tail = _quote_keys_in_chunk(tail)
    tail = tail.replace('undefined', 'null')
    parts.append(tail)

    result = ''.join(parts)

    # Strip trailing commas  ,]  or  ,}  (valid JS, invalid JSON)
    result = re.sub(r',(\s*[}\]])', r'\1', result)

    return result


def strip_const_decl(content):
    """
    Strip  'const testDatabase = '  from start and trailing ';' from end.
    data.js is a single constant declaration so this is safe.
    """
    content = content.strip()
    m = re.match(r'const\s+testDatabase\s*=\s*', content)
    if not m:
        raise ValueError("Cannot find 'const testDatabase =' declaration")
    content = content[m.end():]
    content = content.rstrip()
    if content.endswith(';'):
        content = content[:-1].rstrip()
    return content


# ─── Main ─────────────────────────────────────────────────────────────────────

def main():
    log('=' * 65)
    log('  EXTRACT data.js → per-subject JSON  (v2 fast)')
    log('=' * 65)

    if not os.path.exists(DATA_FILE):
        log(f"[FATAL] {DATA_FILE} not found"); write_log(); sys.exit(1)

    # Step 1: Read
    log(f"\n[STEP 1] Reading {DATA_FILE} ...")
    with open(DATA_FILE, "r", encoding="utf-8") as f:
        content = f.read()
    log(f"         Size: {len(content)/1024/1024:.2f} MB ({len(content):,} chars)")

    # Step 2: Strip const declaration
    log(f"\n[STEP 2] Stripping const declaration ...")
    try:
        js_obj = strip_const_decl(content)
        log(f"         JS object: {len(js_obj)/1024/1024:.2f} MB")
    except Exception as e:
        log(f"[FATAL] {e}"); write_log(); sys.exit(1)

    # Step 3: Convert JS → JSON (regex, fast)
    log(f"\n[STEP 3] Converting JS → JSON (regex-based) ...")
    try:
        json_str = js_to_json(js_obj)
        log(f"         JSON string: {len(json_str)/1024/1024:.2f} MB")
    except Exception as e:
        log(f"[FATAL] Conversion error: {e}")
        log(traceback.format_exc())
        write_log(); sys.exit(1)

    # Step 4: Parse JSON
    log(f"\n[STEP 4] Parsing JSON ...")
    try:
        testDatabase = json.loads(json_str)
    except json.JSONDecodeError as e:
        log(f"[FATAL] JSON parse error — line {e.lineno}, col {e.colno}: {e.msg}")
        ctx_start = max(0, e.pos - 150)
        ctx_end   = min(len(json_str), e.pos + 150)
        log(f"         Context: ...{repr(json_str[ctx_start:ctx_end])}...")
        # Write the bad JSON to a temp file for inspection
        bad_path = "extract_bad_json.tmp"
        with open(bad_path, "w", encoding="utf-8") as f:
            f.write(json_str[max(0, e.pos-2000):min(len(json_str), e.pos+2000)])
        log(f"         Context written to: {bad_path}")
        write_log(); sys.exit(1)

    subject_ids = list(testDatabase.keys())
    log(f"         Subjects found: {', '.join(subject_ids)}")
    log(f"         Total: {len(subject_ids)} subjects")

    # Step 5: Create output directory
    log(f"\n[STEP 5] Creating {OUTPUT_DIR}/ ...")
    os.makedirs(OUTPUT_DIR, exist_ok=True)
    log(f"         OK: {os.path.abspath(OUTPUT_DIR)}")

    # Step 6: Write per-subject JSON files
    log(f"\n[STEP 6] Writing subject files ...\n")
    manifest = {"subjects": {}}
    total_chapters = 0

    for subject_id in subject_ids:
        subject_data = testDatabase[subject_id]
        chapters = sorted(
            subject_data.keys(),
            key=lambda x: int(x) if x.isdigit() else x
        )

        # Get subject name
        subject_name = subject_id
        for ch in chapters:
            d = subject_data.get(ch)
            if isinstance(d, dict) and d.get("subjectName"):
                subject_name = d["subjectName"]
                break

        # Counts
        mcq_count   = sum(len((subject_data.get(ch) or {}).get("mcqData",     [])) for ch in chapters)
        short_count = sum(len((subject_data.get(ch) or {}).get("shortCQData", [])) for ch in chapters)
        full_count  = sum(len((subject_data.get(ch) or {}).get("fullCQData",  [])) for ch in chapters)

        out_path = os.path.join(OUTPUT_DIR, f"{subject_id}.json")
        with open(out_path, "w", encoding="utf-8") as f:
            json.dump(subject_data, f, ensure_ascii=False, indent=2)

        size_kb  = os.path.getsize(out_path) / 1024
        rel_path = f"{OUTPUT_DIR}/{subject_id}.json"

        log(
            f"  ✓ {subject_id:<8}  {len(chapters):>2} ch  "
            f"MCQ:{mcq_count:<5}  sCQ:{short_count:<5}  fCQ:{full_count:<5}"
            f"  {rel_path} ({size_kb:.0f} KB)"
        )

        manifest["subjects"][subject_id] = {
            "name":     subject_name,
            "chapters": chapters,
            "file":     rel_path,
        }
        total_chapters += len(chapters)

    # Step 7: Write manifest.json
    log(f"\n[STEP 7] Writing manifest.json ...")
    manifest_path = os.path.join(OUTPUT_DIR, "manifest.json")
    with open(manifest_path, "w", encoding="utf-8") as f:
        json.dump(manifest, f, ensure_ascii=False, indent=2)
    log(f"         Subjects: {len(subject_ids)},  Chapters: {total_chapters}")

    # Step 8: Validate
    log(f"\n[STEP 8] Validating ...")
    errors = 0
    for sid in subject_ids:
        try:
            with open(os.path.join(OUTPUT_DIR, f"{sid}.json"), "r", encoding="utf-8") as f:
                parsed = json.load(f)
            if not parsed:
                log(f"  ✗ {sid}: empty"); errors += 1
        except Exception as e:
            log(f"  ✗ {sid}: {e}"); errors += 1

    if errors == 0:
        log(f"  ✓ All {len(subject_ids)} files valid")

    # Done
    log("\n" + "=" * 65)
    log(f"  DONE — Subjects: {len(subject_ids)},  Chapters: {total_chapters}")
    log(f"  Output : {os.path.abspath(OUTPUT_DIR)}/")
    log(f"  data.js: UNTOUCHED  ({os.path.getsize(DATA_FILE):,} bytes)")
    log("=" * 65)
    write_log()
    sys.exit(1 if errors > 0 else 0)


if __name__ == "__main__":
    try:
        main()
    except Exception as ex:
        log(f"[EXCEPTION] {ex}")
        log(traceback.format_exc())
        write_log()
        sys.exit(1)

File: testpaper/extract_datajs.js
#!/usr/bin/env node
/**
 * extract_datajs.js
 * =================
 * Reads data.js, evaluates it to obtain testDatabase,
 * then writes one JSON file per subject to static-data/.
 * Also writes static-data/manifest.json.
 *
 * Usage:  node extract_datajs.js
 * Output: static-data/<subject>.json  (one per subject)
 *         static-data/manifest.json
 */

'use strict';

const fs   = require('fs');
const path = require('path');

const DATA_JS    = path.resolve(__dirname, 'data.js');
const OUTPUT_DIR = path.resolve(__dirname, 'static-data');
const LOG_FILE   = path.resolve(__dirname, 'extract_datajs.log');

const logLines = [];
function log(msg) {
    console.log(msg);
    logLines.push(msg);
}

function writeLog() {
    fs.writeFileSync(LOG_FILE, logLines.join('\n') + '\n', 'utf8');
}

// ─── 1. Load data.js ─────────────────────────────────────────────────────────

log('='.repeat(65));
log('  EXTRACT data.js → per-subject JSON');
log('='.repeat(65));

if (!fs.existsSync(DATA_JS)) {
    log('[FATAL] data.js not found at: ' + DATA_JS);
    writeLog();
    process.exit(1);
}

log('\n[STEP 1] Reading data.js ...');
const rawContent = fs.readFileSync(DATA_JS, 'utf8');
log(`         Size: ${(rawContent.length / 1024 / 1024).toFixed(2)} MB`);

// ─── 2. Evaluate to extract testDatabase ─────────────────────────────────────

log('\n[STEP 2] Evaluating data.js ...');
let testDatabase;
try {
    // Use the same proven pattern from scratch/inspect_bus2_datajs.js
    testDatabase = (new Function(rawContent + '\nreturn testDatabase;'))();
} catch (err) {
    log('[FATAL] Failed to evaluate data.js: ' + err.message);
    writeLog();
    process.exit(1);
}

if (!testDatabase || typeof testDatabase !== 'object') {
    log('[FATAL] testDatabase is not a valid object after eval.');
    writeLog();
    process.exit(1);
}

const subjectIds = Object.keys(testDatabase);
log(`         Subjects found: ${subjectIds.join(', ')}`);
log(`         Total subjects: ${subjectIds.length}`);

// ─── 3. Create output directory ───────────────────────────────────────────────

log('\n[STEP 3] Creating static-data/ directory ...');
fs.mkdirSync(OUTPUT_DIR, { recursive: true });
log(`         Output dir: ${OUTPUT_DIR}`);

// ─── 4. Write per-subject JSON files ─────────────────────────────────────────

log('\n[STEP 4] Extracting subjects ...\n');

const manifest = { subjects: {} };
let totalChapters = 0;

for (const subjectId of subjectIds) {
    const subjectData = testDatabase[subjectId];
    const chapters    = Object.keys(subjectData).sort((a, b) => parseInt(a) - parseInt(b));

    // Get human-readable subject name from first chapter that has it
    let subjectName = subjectId;
    for (const ch of chapters) {
        if (subjectData[ch] && subjectData[ch].subjectName) {
            subjectName = subjectData[ch].subjectName;
            break;
        }
    }

    // Count MCQs, shortCQs, fullCQs across all chapters
    let mcqCount = 0, shortCount = 0, fullCount = 0;
    for (const ch of chapters) {
        const d = subjectData[ch];
        if (!d) continue;
        mcqCount   += (d.mcqData      || []).length;
        shortCount += (d.shortCQData  || []).length;
        fullCount  += (d.fullCQData   || []).length;
    }

    const outFile = path.join(OUTPUT_DIR, `${subjectId}.json`);
    const json    = JSON.stringify(subjectData, null, 2);
    fs.writeFileSync(outFile, json, 'utf8');

    const fileSizeKB = (fs.statSync(outFile).size / 1024).toFixed(0);
    const relPath    = `static-data/${subjectId}.json`;

    log(
        `  ✓ ${subjectId.padEnd(8)}` +
        `  ${chapters.length} ch` +
        `  MCQ:${mcqCount}  shortCQ:${shortCount}  fullCQ:${fullCount}` +
        `  → ${relPath} (${fileSizeKB} KB)`
    );

    manifest.subjects[subjectId] = {
        name:     subjectName,
        chapters: chapters,
        file:     relPath,
    };

    totalChapters += chapters.length;
}

// ─── 5. Write manifest.json ───────────────────────────────────────────────────

log('\n[STEP 5] Writing manifest.json ...');
const manifestPath = path.join(OUTPUT_DIR, 'manifest.json');
fs.writeFileSync(manifestPath, JSON.stringify(manifest, null, 2), 'utf8');
log(`         Subjects: ${subjectIds.length}`);
log(`         Chapters: ${totalChapters}`);
log(`         Path: ${manifestPath}`);

// ─── 6. Validation ────────────────────────────────────────────────────────────

log('\n[STEP 6] Validating outputs ...');
let errors = 0;
for (const subjectId of subjectIds) {
    const outFile = path.join(OUTPUT_DIR, `${subjectId}.json`);
    try {
        const parsed = JSON.parse(fs.readFileSync(outFile, 'utf8'));
        const chs = Object.keys(parsed);
        if (chs.length === 0) {
            log(`  ✗ ${subjectId}: empty object`);
            errors++;
        }
    } catch (e) {
        log(`  ✗ ${subjectId}: JSON parse error — ${e.message}`);
        errors++;
    }
}
if (errors === 0) {
    log(`  ✓ All ${subjectIds.length} subject files validated successfully`);
} else {
    log(`  ✗ ${errors} validation error(s) found`);
}

// ─── Done ─────────────────────────────────────────────────────────────────────

log('\n' + '='.repeat(65));
log(`  EXTRACTION COMPLETE`);
log(`  Subjects : ${subjectIds.length}`);
log(`  Chapters : ${totalChapters}`);
log(`  Files    : ${path.join(OUTPUT_DIR, '*.json')}`);
log(`  Original data.js is UNTOUCHED.`);
log('='.repeat(65));

writeLog();
process.exit(errors > 0 ? 1 : 0);
